R/getYearsInBioc.R

```diff
@@ -0,0 +1,205 @@
+#'
+#' @description This function determines the number of years a package has been in
+#'     Bioconductor. Available information includes first Bioconductor version a
+#'     package appeared and the current length of time in Bioconductor. If a
+#'     package has been removed from Bioconductor, information on the last
+#'     Bioconductor version and approximate time in Bioconductor before removal
+#'     is available.
+#' 
+#' @param pkglist List of packages to retrieve information. If default NULL,
+#'     returns a tibble of all Bioconductor packages.
+#' 
+#' @author Lori Shepherd Kern, Robert Shear
+#' 
+#' @return 'tibble' with the following columns:
+#'   * package: name of Bioconductor package
+#'   * category: bioc, data/experiment, data/annotation, workflow
+#'   * first_version_available: Bioconductor version (e.g. 1.9, 3.21) the
+#'     package first became available
+#'   * first_version_release_date: Equivalent calendar date of given
+#'     Bioconductor release
+#'   * approx_years_in: Numeric indicator of years in Bioconductor. If empty,
+#'     indicates package was removed. See final three columns for more
+#'     information.
+#'   * last_version_available: If package was removed from Bioconductor, the
+#'     last Bioconductor version (e.g. 1.9, 3.21) the package was able to be installed
+#'   * last_version_release_date: Equivalent calendar date of given Bioconductor
+#'     release
+#'   * years_before_rm: If removed, how many years it was in Bioconductor
+#' 
+#' @importFrom yaml read_yaml
+#' @importFrom dplyr filter mutate select group_by summarize arrange
+#' @importFrom purrr list_rbind map2
+#' @importFrom glue glue
+#' @importFrom curl curl has_internet
+#' @importFrom tidyr tibble unnest_longer
+#' @importFrom lubridate mdy
+#' @importFrom BiocFileCache bfcquery bfcnew bfcrpath
+#'
+#' @examples
+#' \dontrun{
+#'    ## full table all Bioconductor packages
+#'    tbl <- getPkgYearsInBioc()
+#'
+#'    ## example of package list. Packages active in Bioconductor
+#'    tbl <- getPkgYearsInBioc(c("BiocFileCache", "BiocPkgTools"))
+#'
+#'    ## example of a package that has been removed from Bioconductor
+#'    tbl <- getPkgYearsInBioc("ensemblVEP")
+#' }
+#' 
+#' @export
+
+getPkgYearsInBioc <- function(pkglist=NULL){
+
+    bfc <- .get_cache()
+    rid <- BiocFileCache::bfcquery(bfc, "bioc.pkg.years.info", "rname", exact = TRUE)$rid
+    if (!length(rid)){
+        pkginfo <- BiocFileCache::bfcnew(bfc, "bioc.pkg.years.info", ext = ".Rdata")
+    }else{
+        pkginfo <- BiocFileCache::bfcrpath(bfc, rids = rid)
+    }
+    
+    if (!file.exists(pkginfo)) {
+        NeededYearsData <- .generateBiocYearsInData(pkginfo)
+    }else{       
+        load(pkginfo)
+        if(has_internet()){
+            message("Checking for Bioc Release Update")
+            NeededYearsData <- .updateYearsInBioc(NeededYearsData, pkginfo)
+        }else{
+            message("Cannot check for release update without internet.")
```

Comment:
This should be a `stop` or the `NeededYearsData` variable should be created for the function to continue.